File: app/src/main/java/com/wavesplatform/wallet/v1/crypto/CryptoProvider.java
package com.wavesplatform.wallet.v1.crypto;

import org.whispersystems.curve25519.OpportunisticCurve25519Provider;

import java.lang.reflect.Constructor;

import static com.wavesplatform.wallet.v1.request.TransferTransactionRequest.SignatureLength;

public class CryptoProvider {

    private static OpportunisticCurve25519Provider provider;
    public static OpportunisticCurve25519Provider get() {
        if (provider == null) {
            Constructor<OpportunisticCurve25519Provider> constructor = null;
            try {
                constructor =OpportunisticCurve25519Provider.class.getDeclaredConstructor();
                constructor.setAccessible(true);
                provider = constructor.newInstance();
            } catch (Exception e) {
                e.printStackTrace();
                throw new RuntimeException("Couldn't create crypto provider", e);
            }

        }
        return provider;
    }

    public static byte[] sign(byte[] key, byte[] msg) {
        return get().calculateSignature(CryptoProvider.get().getRandom(SignatureLength), key, msg);
    }
}

File: wavesplatform/src/main/java/com/wavesplatform/sdk/crypto/CryptoProvider.kt
package com.wavesplatform.sdk.crypto

import org.whispersystems.curve25519.OpportunisticCurve25519Provider

import java.lang.reflect.Constructor

internal object CryptoProvider {

    private var SIGNATURE_LENGTH = 64

    private var provider: OpportunisticCurve25519Provider? = null
    fun get(): OpportunisticCurve25519Provider {
        if (provider == null) {
            val constructor: Constructor<OpportunisticCurve25519Provider>
            try {
                constructor = OpportunisticCurve25519Provider::class.java.getDeclaredConstructor()
                constructor.isAccessible = true
                provider = constructor.newInstance()
            } catch (e: Exception) {
                e.printStackTrace()
                throw RuntimeException("Couldn't create crypto provider", e)
            }
        }
        return provider!!
    }

    @JvmStatic
    fun sign(key: ByteArray, msg: ByteArray): ByteArray {
        return get().calculateSignature(CryptoProvider.get().getRandom(SIGNATURE_LENGTH), key, msg)
    }
}
